### ISSUE TITLE
Xcode settings for `PBXCopyFilesBuildPhase`

### ISSUE THREAD

**Issue by KageKirin** [bkaradzic/GENie#394]
Hi,
I'm looking for a way to integrate several PBXCopyFilesBuildPhase into my generated Xcode application projects.

Ideally, I'd like to piggy-back on `resources {}`, maybe in the same way as files can be handled:
```
resources {
 ['targetFolderA'] = { "**.target.bin" ,},
 ['targetFolderB'] = { "**.other-targets ,},
}
```

I see that ~~#265~~ #39 is going into the same direction. What's the ETA on this feature?
And more importantly is there maybe already to do this?

Note that currently I'm just copying my assets as a post-build step calling rsync to copy the assets.

Cheers.

EDIT: I referenced the wrong issue above. #39 being from 2016, I wondered if those changed had already been merged.

**Comment by bkaradzic** [comment]
There is no ETA.

This sounds like it could be pre/post build process command https://github.com/bkaradzic/GENie/blob/master/docs/scripting-reference.md#prebuildcommandscommands. Did you try that?

**Comment by KageKirin** [comment]
`postbuildcommand` is what I'm using.

In case of Xcode, having something like the `resources {}` tag above generate `PBXCopyFilesBuildPhase` would be welcome.
```
6A6C3E7B20D0EAAA00A25248 /* CopyFiles */ = {
			isa = PBXCopyFilesBuildPhase;
			buildActionMask = 2147483647;
			dstPath = Assets;
			dstSubfolderSpec = 6;
			files = (
				6A6C3E7C20D0EAE800A25248 /* engine-assets in CopyFiles */,
				6A6C3E7D20D0EAE800A25248 /* assets in CopyFiles */,
			);
			runOnlyForDeploymentPostprocessing = 0;
		};
```

I.e. I guess, I'll add this feature then.

**Comment by KageKirin** [comment]
More on this: apparently (with latest Xcode), I need to copy my resources through a `PBXCopyFilesBuildPhase` so they can be signed.
I.e. my project merrily copying resources through a post-build `rsync` resulted in an error at `codesign`.

**Comment by KageKirin** [comment]
Upate on the above:

on iOS apps, assets can go directly to a subfolder of the app-root;
on OSX apps, assets must go into a subfolder of app-root/Contents.
Copying assets accordingly fixes the codesign errors I had.

That said, copying frameworks into an app is not as easy as copying them to app-root/Frameworks (iOS) or app-root/Contents/Frameworks (OSX): they need to be re-codesigned.

The usual approach for this is to have a PBXCopyFilesBuildPhase with `dstSubfolderSpec = 10` and the `sign on copy` flag enabled.
There is a workaround with the `carthage copy-frameworks` script as PBXShellScriptBuildPhase, but it requires files to be specified.

Long story short, I'm currently trying to add this to GENie in one of the 2 ways since I really need it to finalize my app pipeline.
Eventually what I want is the following:

- for simple resources
```
resources {
 ['targetFolderA'] = { "**.target.bin" ,},
 ['targetFolderB'] = { "**.other-targets ,},
}
```

- for frameworks (via copy)
```
frameworks {
  path.join(somewhere, "stuff.framework"),
}
```

- for shell commands with input files
```
frameworks {
  ["command"] = { -- files...
    path.join(somewhere, "stuff.framework"),
  }
}
```

**Comment by KageKirin** [comment]
I need a bit of help on this:
So far I added the commands to `api.lua`:
```
		xcodescriptphases =
		{
			kind  = "keypath",
			scope = "config",
		},
		resources =
		{
			kind  = "keypath",
			scope = "config",
		},
		frameworks =
		{
			kind  = "list",
			scope = "config",
		},
```

in `xcode10.project, I call a new function, e.g. `xcode.PBXShellScriptBuildPhase_xcodescriptphases(tr)`

My question now is, inside `xcode.PBXShellScriptBuildPhase_xcodescriptphases()`, I iterate the configs and try to access the corresponding setting, but it seems to be empty.
Do I need to do something else to copy/merge the settings?
```
for _, cfg in ipairs(tr.configs) do
			local cfgcmds = cfg['xcodescriptphases']
``` 

Note: my project has this:
```
xcodescriptphase {
			['hoge ${FOOBAR}'] = {
				path.join(engine_root, "engine-config", "**.config"),
			},
			['fuga ${FOOBAR}'] = {
				path.join(game_root, "config", "**.config"),
			}
		}
```

**Comment by bkaradzic** [comment]
> I need a bit of help on this:

Probably is the best to setup some Lua debugger (https://github.com/bkaradzic/GENie#debugging-genie-scripts) and figure out exactly where data that you need is stored.

**Comment by KageKirin** [comment]
Good advice. Thanks!

**Comment by bkaradzic** [comment]
Also if you prefer printf style debugging, there is printtable (might be easier to start with that, in case you can narrow down quickly where data is going): https://github.com/bkaradzic/GENie/blob/457a46f1187162e201aab944f0a8c71037d3808a/src/base/inspect.lua#L285

**Comment by bkaradzic** [comment]
This `printtable` is accessible from any .lua script.

**Comment by KageKirin** [comment]
Thanks. I'm slowly getting there, with a mix of debugger and printing.

I might end up refactoring `xcode.PBXShellScriptBuildPhase()`'s `doblock()` to avoid duplicating code.
Would it be acceptable to generate 1+ `PBXShellScriptBuildPhase` per configuration (with the necessary `if [ "${CONFIGURATION}" = "' ...]`) to allow different files to be copied per config? Or would this be a regression?

**Comment by bkaradzic** [comment]
That's more Q for @rhoot.

**Comment by KageKirin** [comment]
I'll go and push the current changes to a branch on my fork and create a pull request. We can decide then.

In fact, the more I iterate on this, the more I keep running into a TMTOWTDI problem:  
```
		xcodescriptphases {
			['hoge ${FOOBAR}'] = {
				path.join(engine_root, "engine-config", "**.config"),
			},
		}
```
given  this, the paths I pass to the command (`hoge`) can be passed "as-is"
or I can unroll them into files using `os.matchfiles`
or folders using `os.matchdirs`.
All 3 possibilities seem valid.
Duplicating `xcodescriptphases` into 3 types is even worse, and adding some kind of tag to the path 
(e.g. `[cmd] = { { pathA, 'raw'} ,{ pathB, 'files'}, {pathC, 'dirs'},}`) is mkaing it even more complicated...
Any ideas/preferences on this?

**Comment by rhoot** [comment]
I'm currently between flights on a really long trip, so it will be a couple of days before I've both landed and had my brain has caught up to what's happening, before I'll catch up on this. Sorry for the wait.

**Comment by KageKirin** [comment]
No problem. It's a long weekend in Japan anyway, so it won't be before Tuesday that I get anything in a pushable state.

**Comment by KageKirin** [comment]
Now that you mention that, I didn't think about the key/value-dict issue.
I'll be refactoring my code in the direction of your advice tomorrow.


(How do I make sure the list only contains tuples of 3 elements?)

**Comment by KageKirin** [comment]
Cross-reference to pull request: #397

**Comment by KageKirin** [comment]
Pull request #398 is the script based solution.
#397 will get more changes.

**Comment by KageKirin** [comment]
@bkaradzic @rhoot 
I've run into an interesting limitation in Xcode: PBXScriptPhases and PBXCopyFilesBuildPhases only exist on a _per-project_ basis, not _per-configuration_.
For the script phases, the _per-configuration_ can be handled from inside the script, as it's already done for `prebuildcommands` et al. However, there's no way to force copy phases to be executed per configuration, so I've decided to actually split the API into per-config and per-project settings:
- for the per-config settings I'll use script phases. Pull request #398 will be updated accordingly
- for the per-project settings, I'll use the actual copy phases. Pull request #397 is going to get a bit bigger than I initially expected, as there a few changes in `xcode.getbuildcategory()` and file handling.

note: #398 also has a smell at the moment, in that it is relying on `carthage` to be installed.

It would be tremendously helpful if you could have a look at the pull requests and comment in things to improve.

Cheers.

**Comment by KageKirin** [comment]
so,
#399 adds the per-configuration `xcodescriptphases` keyword

https://github.com/KageKirin/GENie/tree/added-xcodecopyresources adds the `xcodecopyresources` keyword and related functionality. It's based on #399's branch, so I need to rebase this branch once the former has been merged and the I can submit the PR.

https://github.com/KageKirin/GENie/tree/added-xcodecopyframeworks add the `xcodecopyframeworks` keyword. It's also a continuation of `added-xcodecopyresources`, so again, I'll submit PR once its predecessor has been merged.

**Comment by KageKirin** [comment]
I just submitted #406 as the last part of this issue.

**Comment by KageKirin** [comment]
Thanks for reviewing and merging my pull requests.
I can consider this issue solved now.
Cheers!
